core/libraries/messages/EE_Messages_Generator.lib.php

```diff
@@ -961,8 +960,17 @@ public function create_and_add_message_to_queue(EE_Message_To_Generate $message_
 
         $message = $message_to_generate->get_EE_Message();
 
-        // is there a GRP_ID in the request?
-        if ($GRP_ID = EE_Registry::instance()->REQ->get('GRP_ID')) {
+        $GRP_ID = EE_Registry::instance()->REQ->get('GRP_ID', 0) > 0
+            ? EE_Registry::instance()->REQ->get('GRP_ID')
+            : apply_filters(
+                'FHEE__EE_Messages_Generator__create_and_add_message_to_queue_GRP_ID',
+                0,
+                $message,
+                $message_to_generate,
+                $test_send
+            );
```

Comment:
couple things:

 - this should be converted over to use the new Request class
 - you're potentially calling that `get()` method twice
 - you're **only** filtering the `GRP_ID` if one isn't already set??!?! is that correct? ie: why not just filter the `GRP_ID` regardless of its value and let the callback decide?

imho this would be better if it were simply:

```php
/** @var \EventEspresso\core\services\request\RequestInterface $request */
$request = LoaderFactory::getLoader()->getShared('EventEspresso\core\services\request\RequestInterface');
$GRP_ID = apply_filters(
    'FHEE__EE_Messages_Generator__create_and_add_message_to_queue_GRP_ID',
    $request->getRequestParam('GRP_ID', 0),
    $message,
    $message_to_generate,
    $test_send
);
```

then any callbacks hooking into that can decide whether to filter the `GRP_ID` even if it isn't `0`, but even if you don't want to do that, at least change this to be:


```php
/** @var \EventEspresso\core\services\request\RequestInterface $request */
$request = LoaderFactory::getLoader()->getShared('EventEspresso\core\services\request\RequestInterface');
$GRP_ID = $request->getRequestParam('GRP_ID', 0);
$GRP_ID = $GRP_ID > 0
    ? $GRP_ID
    : apply_filters(
        'FHEE__EE_Messages_Generator__create_and_add_message_to_queue_GRP_ID',
        0,
        $message,
        $message_to_generate,
        $test_send
    );
```

Reply:
The original intent was to have any GRP_ID set via the request be explicit because that generally is already an explicit template request.  The filter would only be for when there is no GRP_ID in the request to allow _setting_ an explicit GRP_ID.  Normally, at this point in the stack trace the only time there is a GRP_ID is when it's set via the request.

With that said, after thinking about it, I think it'll be more flexible if we actually filter everything because that way consuming code _can_ override the `GRP_ID` from the request if desired.  It'd still be possible for the client code to derive whether the GRP_ID comes from the request and leave it alone if desired.

I also set the default to be the group id from the `EE_Message` instance on the off chance in the future we start setting `EE_Message::GRP_ID` earlier.  Currently it'll just return `0` if its not set so effectively the same as what I had before.